File: ExpandZeros/ExpandZeros.java

/**
 * Given a matrix, for any element with value 0,
 * set all elements from the corresponding row and column on 0, as well.
 * 
 * Multithreading implementation for Java 1.8.
 * 
 * @2017 Florin Tulba (florintulba@yahoo.com)
 * 
 */

import java.util.ArrayList;
import java.util.Arrays;
import java.util.BitSet;
import java.util.Random;

/**
 * A pair of values: the start index of a range and the past last index
 */
class Range {
	private int startIdx;
	private int pastLastIdx;

	public Range(int startIdx_, int pastLastIdx_) {
		startIdx = startIdx_;
		pastLastIdx = pastLastIdx_;
	}

	public int start() {
		return startIdx;
	}

	public int pastLast() {
		return pastLastIdx;
	}
}

/**
 * Implements the algorithm to expand the zeros and provides code for testing
 * it.
 */
public class ExpandZeros {
	protected static final int processorsCount = Runtime.getRuntime().availableProcessors();

	// generator of random values
	protected static final Random randGen = new Random(System.nanoTime());

	protected static final int TIMES = 1000; // iterations count

	// Max matrix dimensions
	protected static final int mMax = 500; // max number of rows
	protected static final int nMax = 500; // max number of columns
	protected static final int dimAMax = mMax * nMax; // max number of elements
	protected static final short origZerosPercentage = 4; // Desired percentage
															// of zeros within
															// generated
															// matrices

	protected int[] a; // the linearized matrix, preallocated for max allowed
						// size
	protected int m; // the actual number of rows
	protected int n; // the actual number of columns
	protected int dim; // actual number of matrix elements

	protected BitSet checkRows; // the correct rows containing value 0
	protected BitSet checkCols; // the correct columns containing value 0

	protected BitSet foundRows; // the found rows containing value 0
	protected BitSet foundCols; // the found columns containing value 0

	/**
	 * It only preallocates the matrix for max allowed size
	 */
	public ExpandZeros() {
		a = new int[dimAMax];

		m = n = dim = 0;
		checkCols = checkRows = foundCols = foundRows = null;
	}

	/**
	 * Chooses the dimensions and the elements of matrix a and provides
	 * checkRows and checkCols
	 */
	protected void randInitMat() {
		m = randGen.nextInt(mMax - 15) + 15; // random between [15, mMax)
		n = randGen.nextInt(nMax - 15) + 15; // random between [15, nMax)
		dim = m * n;
		checkRows = new BitSet(m);
		checkCols = new BitSet(n);

		// Set all elements first on non-zero values
		for (int i = 0; i < dim; ++i)
			a[i] = randGen.nextInt(999) + 1; // random between [1, 1000)

		// Choose some coordinates to be set to 0
		final int zerosCount = dim * origZerosPercentage / 100;
		for (int i = 0; i < zerosCount; ++i) {
			final int pos = randGen.nextInt(dim), row = pos / n, col = pos % n;
			a[pos] = 0;
			checkRows.set(row);
			checkCols.set(col);
		}
	}

	/**
	 * Signals any mismatches between (checkCols - foundCols) and (checkRows -
	 * foundRows)
	 * 
	 * @return true if there are no mismatches
	 */
	protected boolean correct() {
		if (null == checkRows || null == checkCols || null == foundRows || null == foundCols)
			return true;

		if (!checkRows.equals(foundRows)) {
			System.err.println("Rows Mismatches:");
			System.err.println("\tExpected: " + checkRows);
			System.err.println("\tReceived: " + foundRows);
			return false;
		}

		if (!checkCols.equals(foundCols)) {
			System.err.println("Columns Mismatches:");
			System.err.println("\tExpected: " + checkCols);
			System.err.println("\tReceived: " + foundCols);
			return false;
		}

		return true;
	}

	/**
	 * Traverses `foundCols`and merges all consecutive columns containing true.
	 * So it creates a set of ranges as explained for type `Range`
	 */
	protected ArrayList<Range> buildColRanges() {
		assert foundCols != null;

		ArrayList<Range> result = new ArrayList<>();
		for (int c = 0; c < n; ++c) {
			if (foundCols.get(c)) {
				// Find how many consecutive columns (after c) need to be set on
				// 0
				int c1 = c + 1;
				for (; c1 < n && foundCols.get(c1); ++c1)
					;
				result.add(new Range(c, c1));
				c = c1;
			}
		}
		return result;
	}

	/**
	 * Detects the rows and columns containing the value 0 within the provided
	 * chunk of data
	 */
	protected void findZerosInChunk(int startRow, int pastLastRow) {
		// Avoids false sharing of foundCols through a local vector
		BitSet localFoundCols = new BitSet(n);

		for (int r = startRow; r < pastLastRow; ++r) {
			boolean rowContains0 = false;
			int idx = r * n;
			for (int c = 0; c < n; ++c) {
				if (a[idx++] == 0) {
					localFoundCols.set(c);
					rowContains0 = true;
				}
			}
			if (rowContains0)
				foundRows.set(r); // setting this outside the inner loop
									// minimizes false sharing of foundRows
		}

		// Perform an union of all localFoundCols into foundCols
		for (int c = 0; c < n; ++c) {
			if (localFoundCols.get(c) && !foundCols.get(c)) {
				// Minimizes false sharing of foundCols by reducing writing
				// operations.
				// Overwriting events actually don't change the data,
				// but they invalidate the corresponding L1 cache line from
				// other threads.
				// This is however cheaper than a synchronization mechanism.
				foundCols.set(c);
			}
		}
	}

	/**
	 * Thread detecting the rows and columns containing the value 0 within the
	 * provided chunk of data
	 */
	protected class ZerosDetector extends Thread {
		protected int startRow; // first row to analyze
		protected int pastLastRow; // the index after the last row to analyze

		public ZerosDetector(int startRow_, int pastLastRow_) {
			startRow = startRow_;
			pastLastRow = pastLastRow_;
		}

		@Override
		public void run() {
			findZerosInChunk(startRow, pastLastRow);
		}
	}

	/**
	 * Expands the zeros from the rows and columns containing an original 0 in
	 * the provided chunk of data
	 */
	protected void expandZerosInChunk(ArrayList<Range> colRanges, int startRow, int pastLastRow) {
		for (int r = startRow; r < pastLastRow; ++r) {
			final int rowStart = r * n;

			// Not using the merge of consecutive rows containing value 0
			// since the merge might cover rows tackled by a different thread
			if (foundRows.get(r)) {
				Arrays.fill(a, rowStart, rowStart + n, 0);
			} else {
				for (Range colRange : colRanges) {
					final int startCol = rowStart + colRange.start();
					final int pastLastCol = rowStart + colRange.pastLast();
					Arrays.fill(a, startCol, pastLastCol, 0);
				}
			}
		}
	}

	/**
	 * Thread expanding the zeros from the rows and columns containing an
	 * original 0 in the provided chunk of data
	 */
	protected class ZerosExpander extends Thread {
		protected ArrayList<Range> colRanges; // which columns need to be set on
												// 0 on each row
		protected int startRow; // first row to analyze
		protected int pastLastRow; // the index after the last row to analyze

		public ZerosExpander(ArrayList<Range> colRanges_, int startRow_, int pastLastRow_) {
			colRanges = colRanges_;
			startRow = startRow_;
			pastLastRow = pastLastRow_;
		}

		@Override
		public void run() {
			expandZerosInChunk(colRanges, startRow, pastLastRow);
		}
	}

	/**
	 * Expands all the zeros from the matrix on the corresponding rows and
	 * columns
	 */
	public void apply() {
		assert null != a && m > 0 && n > 0 && dim == m * n && checkRows != null && checkCols != null;

		foundRows = new BitSet(m);
		foundCols = new BitSet(n);

		ArrayList<Thread> threads = new ArrayList<>();

		// Distribute several consecutive rows to every spawn thread and keep
		// the remaining ones to be processed in this main thread.

		// There is no point spawning a thread unless it has at least 300000
		// elements to analyze
		// The main thread should process most of the time fewer elements, as it
		// also needs to wait for the created threads
		int chunkSz = (int) Math.ceil(Math.max((double) m / processorsCount, 300000. / n));

		int chunkStartRow = 0, chunkPastLastRow = chunkSz;
		int additionalThreads = (int) Math.ceil(m / chunkSz) - 1;
		while (additionalThreads-- > 0) {
			threads.add(new ZerosDetector(chunkStartRow, chunkPastLastRow));
			chunkStartRow = chunkPastLastRow;
			chunkPastLastRow += chunkSz;
		}

		for (Thread t : threads)
			t.start();

		findZerosInChunk(chunkStartRow, m);

		for (Thread t : threads)
			for (;;)
				try {
					t.join();
					break; // join next thread
				} catch (InterruptedException e1) {
					continue; // just retry joining t
				}

		final ArrayList<Range> colRanges = buildColRanges();

		chunkStartRow = 0;
		chunkPastLastRow = chunkSz;
		additionalThreads = threads.size();
		while (additionalThreads-- > 0) {
			threads.set(additionalThreads, new ZerosExpander(colRanges, chunkStartRow, chunkPastLastRow));
			chunkStartRow = chunkPastLastRow;
			chunkPastLastRow += chunkSz;
		}

		for (Thread t : threads)
			t.start();

		expandZerosInChunk(colRanges, chunkStartRow, m);

		for (Thread t : threads)
			for (;;)
				try {
					t.join();
					break; // join next thread
				} catch (InterruptedException e1) {
					continue; // just retry joining t
				}
	}

	/**
	 * Launches the tests for this class
	 * 
	 * @param args
	 *            unused for now
	 */
	public static void main(String[] args) {
		ExpandZeros proc = new ExpandZeros();
		long elapsed = 0;
		long lastStart = 0;
		long totalElems = 0;
		for (int i = 0; i < TIMES; ++i) {
			proc.randInitMat();
			totalElems += proc.dim;
			lastStart = System.nanoTime();
			proc.apply();
			elapsed += System.nanoTime() - lastStart;
			assert proc.correct();
		}

		System.out.println("Processed " + totalElems + " elements in " + TIMES + " matrices in " + elapsed / 1e9 + "s");
		System.out.println("That is " + elapsed / (double) totalElems + " ns / element.");
	}
}

File: ExpandZeros/expandZeros.py
#! python3

'''
Given a matrix, for any element with value 0,
set all elements from the corresponding row and column on 0, as well.

Required modules:
 - <bitarray> - https://pypi.python.org/pypi/bitarray

@2017 Florin Tulba (florintulba@yahoo.com)
'''

import sys
import random
import time
import math
import multiprocessing
from itertools import repeat
from bitarray import bitarray
from threading import Thread

# Matrix settings
mMax = 500 # max number of rows
nMax = 500 # max number of columns
dimAMax = mMax * nMax # max number of elements
origZerosPercentage = 4 # Desired percentage of zeros within generated matrices

# Threshold below which a thread should not be created
MinElemsPerThread = 200000 # minimum number of matrix elements to be processed by a created thread
processorsCount = multiprocessing.cpu_count()

class ExpandZeros:
	'''
	Implements the algorithm to expand the zeros and provides code
	for testing it.
	'''
	def __init__(self):
		'''
		It only preallocates the matrix for max allowed size
		'''
		self.a = dimAMax * [0] # the linearized matrix, preallocated for max allowed
		self.m = 0		# the actual number of rows
		self.n = 0		# the actual number of columns
		self.dim = 0	# actual number of matrix elements
		self.checkRows = None # the correct rows containing value 0
		self.checkCols = None	# the correct columns containing value 0
		self.foundRows = None # the found rows containing value 0
		self.foundCols = None # the found columns containing value 0

	def randInitMat(self):
		'''
		Chooses the dimensions and the elements of matrix a and provides
		checkRows and checkCols
		'''
		self.m = random.randrange(15, mMax) # random between [15, mMax)
		self.n = random.randrange(15, nMax) # random between [15, nMax)
		self.dim = self.m * self.n
		self.checkRows = bitarray(self.m * '0')
		self.checkCols = bitarray(self.n * '0')

		# Set all elements first on non-zero values
		for i in range(self.dim):
			self.a[i] = random.randrange(1, 1000) # random between [1, 1000)

		# Choose some coordinates to be set to 0
		zerosCount = self.dim * origZerosPercentage // 100
		for i in range(zerosCount):
			pos = random.randrange(self.dim)
			row, col = pos // self.n, pos % self.n
			self.a[pos] = 0
			self.checkRows[row], self.checkCols[col] = True, True

	def correct(self):
		'''
		Signals any mismatches between (checkCols - foundCols) and (checkRows - foundRows)
		Returns True if there are no mismatches
		'''
		if None == self.checkRows or None == self.checkCols or \
			None == self.foundRows or None == self.foundCols:
			return True

		if self.checkRows != self.foundRows:
			print("Rows Mismatches:", file=sys.stderr)
			print("\tExpected:", self.checkRows, file=sys.stderr)
			print("\tReceived:", self.foundRows, file=sys.stderr)
			return False

		if self.checkCols != self.foundCols:
			print("Columns Mismatches:", file=sys.stderr)
			print("\tExpected: ", self.checkCols, file=sys.stderr)
			print("\tReceived: ", self.foundCols, file=sys.stderr)
			return False

		return True

	def buildColIndices(self):
		'''
		extract the indices of the columns containing value 0
		'''
		colIndices = []
		idx = -1
		while True:
			try: # Searching for the 1st True after idx
				idx = self.foundCols.index(True, idx + 1)
			except ValueError:
				break
			colIndices.append(idx)
		return colIndices

	def findZerosInChunk(self, chunkRowsRange):
		'''
		Detects the rows and columns containing the value 0
		within the provided chunk of data
		'''
		# Avoids false sharing of foundCols through a local vector
		localFoundCols = bitarray(self.n * '0')

		for r in chunkRowsRange:
			rowContains0 = False
			idx = r * self.n
			for c in range(self.n):
				if self.a[idx] == 0:
					localFoundCols[c], rowContains0 = True, True
				idx += 1
			if rowContains0:
				self.foundRows[r] = True

		# Perform an union of all localFoundCols into foundCols
		for c in range(self.n):
			if localFoundCols[c] and not self.foundCols[c]:
				'''
				Minimizes false sharing of foundCols by reducing writing operations.
				Overwriting events actually don't change the data,
				but they invalidate the corresponding L1 cache line from other threads.
				This is however cheaper than a synchronization mechanism.
				'''
				self.foundCols[c] = True

	class ZerosDetector(Thread):
		'''
		Thread detecting the rows and columns containing the value 0
		within the provided chunk of data
		'''
		def __init__(self, outer, chunkRowsRange):
			Thread.__init__(self)
			self.outer = outer
			self.chunkRowsRange = chunkRowsRange # the chunk of data between 2 rows

		def run(self):
			# Override
			self.outer.findZerosInChunk(self.chunkRowsRange)

	def expandZerosInChunk(self, colIndices, chunkRowsRange):
		'''
		Expands the zeros from the rows and columns containing an original 0
		in the provided chunk of data
		'''
		for r in chunkRowsRange:
			rowStart = r * self.n
			if self.foundRows[r]:
				self.a[rowStart : rowStart + self.n] = repeat(0, self.n)
			else:
				for colIdx in colIndices:
					self.a[rowStart + colIdx] = 0

	class ZerosExpander(Thread):
		'''
		Thread expanding the zeros from the rows and columns
		containing an original 0 in the provided chunk of data
		'''
		def __init__(self, outer, colIndices, chunkRowsRange):
			Thread.__init__(self)
			self.outer = outer
			self.colIndices = colIndices # the columns where original zeros were found
			self.chunkRowsRange = chunkRowsRange # the chunk of data between 2 rows

		def run(self):
			# Override
			self.outer.expandZerosInChunk(self.colIndices, self.chunkRowsRange)

	def apply(self):
		'''
		Expands all the zeros from the matrix on the corresponding rows and columns
		'''
		assert [] != self.a and self.m > 0 and self.n > 0 \
			and self.dim == self.m * self.n \
			and self.checkRows != None and self.checkCols != None

		self.foundRows = bitarray(self.m * '0')
		self.foundCols = bitarray(self.n * '0')

		'''
		Distribute several consecutive rows to every spawn thread and
		keep the remaining ones to be processed in this main thread.

		There is no point spawning a thread unless it has a minimum
		number of elements to analyze.
		
		The main thread should process most of the time fewer elements,
		as it also needs to wait for the created threads.
		'''
		chunkSz = int(math.ceil( \
			max(self.m / processorsCount, \
				MinElemsPerThread / self.n)))
		chunkStartRow, chunkPastLastRow = 0, chunkSz
		additionalThreads = int(math.ceil(self.m / chunkSz)) - 1
		threads = []
		while additionalThreads > 0:
			additionalThreads -= 1
			threads.append( \
				ExpandZeros.ZerosDetector(self, \
					range(chunkStartRow, chunkPastLastRow)))
			chunkStartRow = chunkPastLastRow
			chunkPastLastRow += chunkSz
		for t in threads:
			t.start()

		self.findZerosInChunk(range(chunkStartRow, self.m))

		for t in threads:
			t.join()

		colIndices = self.buildColIndices()

		chunkStartRow, chunkPastLastRow = 0, chunkSz
		additionalThreads = len(threads)
	
		while additionalThreads > 0:
			additionalThreads -= 1
			threads[additionalThreads] = \
				ExpandZeros.ZerosExpander(self, \
					colIndices, \
					range(chunkStartRow, chunkPastLastRow))
			chunkStartRow = chunkPastLastRow
			chunkPastLastRow += chunkSz
		for t in threads:
			t.start()

		self.expandZerosInChunk(colIndices, range(chunkStartRow, self.m))

		for t in threads:
			t.join()

if __name__ == "__main__":
	random.seed()
	elapsed, lastStart, totalElems = 0, 0, 0
	proc = ExpandZeros()
	TIMES = 10 # iterations count
	for i in range(TIMES):
		proc.randInitMat()
		totalElems += proc.dim
		lastStart = time.perf_counter()
		proc.apply()
		elapsed += time.perf_counter() - lastStart
		assert proc.correct()
	print("Processed", totalElems, "elements in", TIMES, "matrices in %.3fs" % elapsed);
	print("That is %.3fns / element." % (elapsed * 1e9 / totalElems));
